decimaldate.js

```diff
@@ -9,129 +9,157 @@ decimaldate.DECIMALPLACES = 6;
 decimaldate.RE_YEARMONTHDAY = /^(\-?\+?)(\d+)\-(\d\d)\-(\d\d)$/;
 
 decimaldate.iso2dec = (isodate) => {
+    // parse the date into 3 integers and maybe a minus sign
+    // validate that it's a valid date
     const datepieces = isodate.match(decimaldate.RE_YEARMONTHDAY);
-    if (! datepieces) throw (`Invalid date format ${isodate}`);
-
-    [plusminus, yearstring, monthstring, daystring] = datepieces.slice(1);
-    if (! decimaldate.isvalidmonth(monthstring) || ! decimaldate.isvalidmonthday(yearstring, monthstring, daystring))  throw `Invalid date ${isodate}`;
-
-    let decbit = decimaldate.proportionofdayspassed(yearstring, monthstring, daystring);
-    if (plusminus == '-') decbit = 1 - decbit;
-
-    let yeardecimal = parseInt(yearstring) + decbit;
-    if (plusminus == '-') yeardecimal *= -1;
+    if (! datepieces) throw new Error(`iso2dec() malformed date ${isodate}`);
+
+    const [plusminus, yearstring, monthstring, daystring] = datepieces.slice(1);
+    const monthint = parseInt(monthstring);
+    const dayint = parseInt(daystring);
+    let yearint = plusminus == '-' ? -1 * parseInt(yearstring) : parseInt(yearstring);
+    if (yearint <= 0) yearint -= 1;  // ISO 8601 shift year<=0 by 1, 0=1BCE, -1=2BCE; we want proper negative integer
+    if (! decimaldate.isvalidmonthday(yearint, monthint, dayint))  throw new Error(`iso2dec() invalid date ${isodate}`);
+
+    // number of days passed = decimal portion
+    // if BCE <=0 then count backward from the end of the year, instead of forward from January
+    const decbit = decimaldate.proportionofdayspassed(yearint, monthint, dayint);
+
+    let decimaloutput;
+    if (yearint < 0) {
+        // ISO 8601 shift year<=0 by 1, 0=1BCE, -1=2BCE; we want string version
+        // so it's 1 to get from the artificially-inflated integer (string 0000 => -1 for math, +1 to get back to 0)
+        decimaloutput = 1 + 1 + yearint - (1 - decbit);
+    }
+    else {
+        decimaloutput = yearint + decbit;
+    }
 
-    return parseFloat(yeardecimal.toFixed(decimaldate.DECIMALPLACES));
+    // round to standardized number of decimals
+    decimaloutput = parseFloat(decimaloutput.toFixed(decimaldate.DECIMALPLACES));
+    return decimaloutput;
 };
 
 
 decimaldate.dec2iso = (decdate) => {
-    // strip the integer/year part
-    // find how many days were in this year, multiply back out to get the day-of-year number
-    if (decdate >= 0) {
-        yearint = parseInt(Math.abs(Math.floor(decdate)));
-        plusminus = '';
+    // remove the artificial +1 that we add to make positive dates look intuitive
+    const truedecdate = decdate - 1;
```

Comment:
Juggling two different versions of decimal dates is a red flag. While it might work, it means we’re probably doing some conversion just to convert back. I suspect a lot of this shifting is only necessary for the leap year computation, which is unnecessary if we rely more heavily on `Date`.